zboxcore/sdk/copyworker.go

```diff
@@ -160,14 +164,33 @@ func (req *CopyRequest) ProcessCopy() error {
 			defer wg.Done()
 			refEntity, err := req.copyBlobberObject(req.blobbers[blobberIdx], blobberIdx)
 			if err != nil {
+				select {
+					case wgErrors <- err:
+					default:
+				}
 				l.Logger.Error(err.Error())
 				return
 			}
 			objectTreeRefs[blobberIdx] = refEntity
 		}(int(pos))
 	}
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
```

Comment:
We don't need to put the `wg.Wait()` in goroutine, and use `wgDone` channel. We cam simplify it with:
```go
// for loop to copy object the same as above

wg.Wait()
var err error
select {
case err = <-wgErrors:
default:
}

if !req.isConsensusOk() && err != nil {
  return errors.New("copy_failed", fmt.Sprintf("copy failed: %v", err))
}

// keep remaining the same
```

Reply:
@peterlimg 
If we don't use wgDone channel then
https://github.com/0chain/gosdk/actions/runs/4460231710/jobs/7833235220?pr=847 
- **by using wg.wait() in goroutine:** the go routines will all be completed but no error will be received from the wgErrors channel because imo default will be executed before even the error is pushed in channel.
- **by without using wg.wait() in goroutine:** it will create panic

Is it possible to do so in some other way?

Reply:
@Manish-210 wg.wait() is suppose to run outside of the goroutines. The persuade code would be:

```go
wg := sync.WaitGroup{}
for i := 0; i < n; i++ {
  wg.Add(1)
  go func() {
      defer wg.Done()
      if err := doSomthing(); err != nil {
        select {
        case wgErrors <- err:
        default:
        }
        return
      }
}()
}

wg.Wait() 
// all goroutines have finished
select {
case err := wgErrors:
default:  // it would not got to default case if any error happens in the goroutine above, it is because `wg.Wait()` guarantees that all goroutines have finished, which means the error must have been pushed to channel or ignored if the channel is full.
}
```
Do you mean we will panic if we do it in this way? but how?

Reply:
Thanks @peterlimg for the suggestion
https://go.dev/play/p/PLe31txueLn
I tried to implement the same approach. But it will create a deadlock cause we are listening after the goroutines have been completed wgErrors channel.

Reply:
Try again, we need to use `select ... default` case when pushing err to `wgErrors` channel in case the channel is full.